### PR TITLE
Uncaught exceptions analysis

### PR DESCRIPTION
**Pull request by cristina-david**
This analysis computes for each function an overapproximation of the set of exceptions that escape it.

It is be used by a future PR to reduce the number of exceptional return variables introduced by the exception handling.

### CODE REVIEW

File: src/goto-programs/uncaught_exceptions.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+
+Module: Compute an overapproximation of the set of exceptions 
+        that escape each function
+
+Author: Cristina David
+
+Date:   January 2016
+
+\*******************************************************************/
+// #define DEBUG
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: remove

File: src/goto-programs/uncaught_exceptions.cpp [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+#include <map>
+#include <set>
+
+#include <goto-programs/goto_functions.h>
+#include <analyses/call_graph.h>
+#include <util/ui_message.h>
+#include <util/xml.h>
+#include "uncaught_exceptions.h"
+
+class exceptiont
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: I have seen the same definition in remove_exceptions...

File: src/goto-programs/uncaught_exceptions.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+  catch_handlerst catch_handlers;
+};
+
+typedef std::vector<exceptiont> stack_catcht;
+
+class uncaught_exceptionst
+{
+public:
+  explicit uncaught_exceptionst(
+    symbol_tablet &_symbol_table,
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: Make them const if you don't modify them.

File: src/goto-programs/uncaught_exceptions.cpp [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+      stack_catcht stack_caught;
+      std::set<irep_idt> thrown;
+      goto_programt &goto_program=current_function->second.body;
+
+      if(goto_program.empty())
+        continue;
+
+      Forall_goto_program_instructions(instr_it, goto_program)
+      {
+        if(instr_it->type==THROW)
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: switch?

File: src/goto-programs/uncaught_exceptions.cpp [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+  Inputs:
+
+ Outputs:
+
+ Purpose: computes a map associating each function with 
+          an overapproximation of the set of exceptions that escape it
+
+\*******************************************************************/
+
+void uncaught_exceptionst::collect_uncaught_exceptions(
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: It would be nicer to use ait...

File: src/goto-programs/uncaught_exceptions.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+Date:   January 2016
+
+\*******************************************************************/
+
+#ifndef CPROVER__SRC_GOTO_PROGRAMS_UNCAUGHT_EXCEPTIONS_H
+#define CPROVER__SRC_GOTO_PROGRAMS_UNCAUGHT_EXCEPTIONS_H
+
+typedef std::map<irep_idt, std::set<irep_idt>> exceptions_mapt;
+
+void uncaught_exceptions(symbol_tablet &,
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.h]: cpplint

File: src/util/irep_ids.txt [hunk 7]
```diff
@@ -800,4 +800,5 @@ cprover_string_to_char_array_func
 cprover_string_to_lower_case_func
 cprover_string_to_upper_case_func
 cprover_string_trim_func
-cprover_string_value_of_func
\ No newline at end of file
+cprover_string_value_of_func
```
Comment (peterschrammel) [inline src/util/irep_ids.txt]: what has changed?

File: src/util/irep_ids.txt [hunk 8]
```diff
@@ -800,4 +800,5 @@ cprover_string_to_char_array_func
...
+handler
```
Comment (peterschrammel) [inline src/util/irep_ids.txt]: where is the handler assigned? I only see a use of it here. In which PR is it introduced?

File: src/goto-programs/uncaught_exceptions.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+Module: Compute the set of uncaught exceptions per function
+
+Author: Cristina David
+
+Date:   January 2016
+
+\*******************************************************************/
+
+#ifndef CPROVER__SRC_GOTO_PROGRAMS_UNCAUGHT_EXCEPTIONS_H
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.h]: wrong header guard

File: src/goto-programs/uncaught_exceptions.cpp [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+Outputs:
+
+Purpose:
+
+\*******************************************************************/
+
+void uncaught_exceptionst::operator()(exceptions_mapt &exceptions_map)
+{
+  collect_uncaught_exceptions(exceptions_map);
+#ifdef DEBUG
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: Can you integrate this into goto-instrument and add a few tests?

File: src/goto-programs/uncaught_exceptions.cpp [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+
+Outputs:
+
+Purpose: populates exceptions_map with the uncaught exceptions 
+corresponding to each function
+
+\*******************************************************************/
+
+void uncaught_exceptions(
+  symbol_tablet &symbol_table,
```
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: Make things const if you don't modify them

**Comment by forejtv** [comment]
Is this independent of #506, or does one depend on the other?

**Comment by cristina-david** [comment]
In principle, this PR is not dependent on #506 in the sense that the two can be merged in whatever order. However, conceptually speaking, this one doesn't make sense without #506.

File: src/util/irep_ids.txt (hunk 8 above)
Comment (cristina-david) [inline src/util/irep_ids.txt]: It's introduced in #506. That's where the exception handling instrumentation is being generated.

File: src/util/irep_ids.txt (hunk 7 above)
Comment (cristina-david) [inline src/util/irep_ids.txt]: Nothing intentional, just spacing.

File: src/goto-programs/uncaught_exceptions.cpp (hunk 10 above)
Comment (cristina-david) [inline src/goto-programs/uncaught_exceptions.cpp]: Do you want this in the same PR?

File: src/goto-programs/uncaught_exceptions.cpp (hunk 5 above)
Comment (tautschnig) [inline src/goto-programs/uncaught_exceptions.cpp]: I would have thought so as well; and in theory the escape analysis should already lay some foundations for achieving this?

File: src/goto-programs/uncaught_exceptions.cpp [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+  const symbol_tablet &symbol_table,
+  const goto_functionst &goto_functions,
+  exceptions_mapt &exceptions_map)
+{
+  uncaught_exceptionst uncaught_exceptions(symbol_table,
+                                           goto_functions);
+  uncaught_exceptions(exceptions_map);
+}
+
+
```
Comment (tautschnig) [inline src/goto-programs/uncaught_exceptions.cpp]: Nit picking: blank lines at end-of-file

**Comment by cristina-david** [comment]
@peterschrammel @tautschnig I'll redo this to use the AI framework.

File: src/goto-programs/uncaught_exceptions.cpp (hunk 10 above)
Comment (peterschrammel) [inline src/goto-programs/uncaught_exceptions.cpp]: yes, would be good in order to this PR

**Comment by cristina-david** [comment]
@tautschnig @peterschrammel The ait framework uses goto_program_templatet::get_successors, which doesn't do anything for ```throw``` (as it's difficult to figure out the successors). This is OK for the other analyses as the exception handling instrumentation is meant to eliminate throws. However, the current analysis needs to be performed before the throw elimination as it is used to limit the exception handling instrumentation being introduced (i.e. limit the set of methods for which we introduce exceptional return variables). Thus, ait doesn't seem like the best choice for this analysis. I've however improved my initial design to separate the analysis and the domain (without inheriting from ait though..).

**Comment by tautschnig** [comment]
Thanks a lot for all the rework - could you please address the build failure? ("irep_idt get_static_type(const exprt&)' defined but not used")

**Comment by martin-cs** [comment]
On Tue, 2017-03-14 at 07:51 -0700, cristina-david wrote:
> @tautschnig @peterschrammel The ait framework uses goto_program_templatet::get_successors, which doesn't do anything for ```throw``` (as it's difficult to figure out the successors). This is OK for the other analyses as the exception handling instrumentation is meant to eliminate throws.

Ummmm... might I suggest that this is not a long-term solution.
Function pointer removal means we don't have to worry about function
pointers ... in theory.  One task we have is to make ait et al. able to
handle function pointers as static removal is not enough.  I have a
nasty feeling the same might be true for throw.

>  However, the current analysis needs to be performed before the throw elimination as it is used to limit the exception handling instrumentation being introduced (i.e. limit the set of methods for which we introduce exceptional return variables). Thus, ait doesn't seem like the best choice for this analysis. I've however improved my initial design to separate the analysis and the domain (without inheriting from ait though..).

As one of the people who has been working with ait more recently, would
it help if we had a chat about this?

**Comment by peterschrammel** [comment]
Yes, it's a hen-and-egg problem of requiring an analysis to remove a language feature whose removal would make the analysis itself more uniform.

**Comment by cristina-david** [comment]
@tautschnig I've fixed the build problem. @martin-cs Sure, let's talk. At this point, I do think that simplifying the analysed language is the way to go but I'm curious about the problems you've encountered with this.

**Comment by peterschrammel** [comment]
@martin-cs, @cristina-david, any conclusions on that?

**Comment by martin-cs** [comment]
On Mon, 2017-03-20 at 14:22 -0700, Peter Schrammel wrote:
> @martin-cs, @cristina-david, any conclusions on that?

We haven't had a chance to chat about it yet.

**Comment by tautschnig** [comment]
This appears to require a rebase to drop merge commits.

**Comment by cristina-david** [comment]
I've redone this as #940 based on cbmc/test-gen-support

**Pull request closed without merging**